pkg/controller/daemonsetjob.go

```diff
@@ -0,0 +1,273 @@
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: v1core.New(kubeClient.Core().RESTClient()).Events("")})
+
+	jobInformer := kubeInformerFactory.Batch().V1().Jobs()
+	daemonsetjobInformer := deamonsetJobInformerFactory.Daemonsetjob().V1().DaemonsetJobs()
+
+	dj := &DaemonsetJobController{
+		DaemonsetJobClient: daemonsetjobClient,
+		KubeClient:         kubeClient,
+		DaemonsetJobLister: daemonsetjobInformer.Lister(),
+		DaemonsetJobSynced: daemonsetjobInformer.Informer().HasSynced,
+		JobLister:          jobInformer.Lister(),
+		JobSynced:          jobInformer.Informer().HasSynced,
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "workflow"),
+		config: DaemonsetJobControllerConfig{
+			RemoveInvalidDaemonsetJob: true,
+			NumberOfWorkers:           1,
+		},
+
+		Recorder: eventBroadcaster.NewRecorder(scheme.Scheme, apiv1.EventSource{Component: "workflow-controller"}),
+	}
+	daemonsetjobInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc:    dj.onAddDaemonsetJob,
+			UpdateFunc: dj.onUpdateDaemonsetJob,
+			DeleteFunc: dj.onDeleteDaemonsetJob,
+		},
+	)
+
+	jobInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			//	AddFunc:    dj.onAddJob,
+			//	UpdateFunc: dj.onUpdateJob,
+			//	DeleteFunc: dj.onDeleteJob,
+		},
+	)
+	//	dj.JobControl = &DaemonsetJobJobControl{kubeClient, wc.Recorder}
+	dj.updateHandler = dj.updateDaemonsetJob
+	dj.syncHandler = dj.sync
+
+	return dj
+}
+
+// Run simply runs the controller. Assuming Informers are already running: via factories
+func (d *DaemonsetJobController) Run(ctx context.Context) error {
+	glog.Infof("Starting workflow controller")
+
+	if !cache.WaitForCacheSync(ctx.Done(), d.JobSynced, d.DaemonsetJobSynced) {
+		return fmt.Errorf("Timed out waiting for caches to sync")
+	}
+
+	for i := 0; i < d.config.NumberOfWorkers; i++ {
+		go wait.Until(d.runWorker, time.Second, ctx.Done())
+	}
+
+	<-ctx.Done()
+	return ctx.Err()
+}
+
+func (d *DaemonsetJobController) runWorker() {
+	for d.processNextItem() {
+	}
+}
+
+func (d *DaemonsetJobController) processNextItem() bool {
+	key, quit := d.queue.Get()
+	if quit {
+		return false
+	}
+	defer d.queue.Done(key)
+	err := d.syncHandler(key.(string))
+	if err == nil {
+		d.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("Error syncing daemonsejob: %v", err))
+	d.queue.AddRateLimited(key)
+
+	return true
+}
+
+// enqueue adds key in the controller queue
+func (d *DaemonsetJobController) enqueue(daemonsetjob *dapi.DaemonsetJob) {
+	key, err := cache.MetaNamespaceKeyFunc(daemonsetjob)
+	if err != nil {
+		glog.Errorf("DaemonsetJobController:enqueue: couldn't get key for DaemonsetJob   %s/%s", daemonsetjob.Namespace, daemonsetjob.Name)
+		return
+	}
+	d.queue.Add(key)
+}
+
+// get workflow by key method
+func (d *DaemonsetJobController) getWorkflowByKey(key string) (*dapi.DaemonsetJob, error) {
```

Comment:
Is it intentional to have a Workflow related method in the DaemonSetJob code?  So far it looked like we could upstream this if folks were interested after having proved it here, however this snippet would make that harder I suppose.

Reply:
done